Changed Grass coloring from TFC Seasonal

Lumintorious opened the pull request:
Integrated the addon "TFC Seasonal" into the base TFC:TNG mod.

- Added individual grass noise for less stagnant landscapes. It uses a free to use and distribute OpenSimplexNoise(found in util) library.
- Added seasonal grass coloring. Configurable for July, October, January, April. Other months are weighted blends of those.
- Added configurations for colors/noise. Noise scale, darkness and levels. Seasonal color's individual RGBAs to blend into the base grass color computed by TFC. Everything is able to be enabled/disabled.

Feel free to change the base color values in the config to better suit the original TFC landscape vision.

Lumintorious committed:
Changed Grass coloring from TFC Seasonal

- Added individual grass noise for less stagnant landscapes
- Added seasonal grass coloring
- Added configurations for colors/noise,

eerussianguy commented:
Any idea how this compares to classic's grass coloring? I know people really liked the paler / yellower grasses from there

Lumintorious commented:
Yes. I think mine is more saturated cause I preferred it that way. If you want less saturated colors, just set a default config value for the seasonal colors with less saturation or less alpha. I think colorpicking from the minecraft colormap will yield the closest values. Right now summer is barely changed, autumn has a pretty big impact towards yellow/reddish, winter is rather pale blue and spring is greener than summer.

alcatrazEscapee reviewed (changes requested):
Please use our code style. You've made unnecessary import changes, inconsistent spaces and tabs and bracket convention.

You also added open simplex noise... and then never used it in favor of just using the MC perlin noise (which is perfectly sufficient for this purpose). Remove.

eerussianguy commented:
#1114

Lumintorious committed:
Formatting changes

- Removed unused classes
- Formatted towards consistency

Lumintorious commented:
- Removed unused class OpenSimplexNoise
- Formatted brackets, spaces and naming
- Imports are the minimal amount necessary (deduced by eclipse at least)
- Grass and Foliage use the same color function #1114 (temperature + humidity + month = color)

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+{
+    static NoiseGeneratorPerlin NOISE_GRASS = new NoiseGeneratorPerlin(new Random("NOISE_GRASS".hashCode()), 2);
+    
+	public static Color january, february, march, april, may,
```
This would be much simpler if it was a `Color[]` indexed by month ordinals e.g. january = `monthlyColors[Month.JANUARY.ordinal()]`

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+    	august = blendWithAlphas(july, october, 0.65, 0.35);
+    	november = blendWithAlphas(october, january, 0.65, 0.35);
+    	february = blendWithAlphas(january, april, 0.65, 0.35);
+    	may = blendWithAlphas(april, july, 0.65, 0.35);
```
```java
for (int i = 0; i < 12; i += 3)
{
    monthlyColors[i + 1] = blendWithAlphas(monthlyColors[i], monthlyColors[(i + 3) % 12], 0.35);
    monthlyColors[i + 2] = blendWithAlphas(monthlyColors[i], monthlyColors[(i + 3) % 12], 0.65);
}
```
should do it, may need to double check the math there.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+    
+    public static Color parseColor(String s)
+    {
+    	String[] parts = s.split(",");
```
This does not validate the color is correct, and will throw if it is not (in many ways). Since this is a user config this is a big bad. What should be done instead: just make the config options integers, e.g. `0x1EFF1ECC`. No input validation required, no parsing required.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+        return ColorizerGrass.getGrassColor(0.5, 0.5);
+    }
+    
+    public static Color getSeasonalColor()
```
This method becomes `monthyColors[CalendarTFC.CALENDAR_TIME.getMonthOfYear().ordinal()]`

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+        return newLow + (value - currentLow) * (newHigh - newLow) / (currentHigh - currentLow);
+    }
+	
+	public static double clamp(double value, double min, double max)
```
`MathHelper.clamp()` already exists.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+        amount = clamp(amount, 0, 1);
+        amount = clamp((amount - start) / (end - start), 0, 1);
+        
+        return clamp(0.5 + 0.5 * Math.sin(Math.cos(Math.PI * Math.tan(90 * amount))) * Math.cos(Math.sin(Math.tan(amount))), 0, 1);
```
Use `MathHelper.sin` / `MathHelper.cos`

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+        return clamp(0.5 + 0.5 * Math.sin(Math.cos(Math.PI * Math.tan(90 * amount))) * Math.cos(Math.sin(Math.tan(amount))), 0, 1);
+    }
+	
+	public static Color blendByWeight(Color c0, Color c1, double weight0, double weight1)
```
`weight1` is always `1 - weight0`, inline the param.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+	    return new Color((int) r, (int) g, (int) b, (int) a);
+	}
+	
+	public static Color blendWithAlphas(Color c0, Color c1, double weight0, double weight1)
```
`weight1` is always `1 - weight0`, inline the param.

DisasterMoo commented on src/main/java/net/dries007/tfc/ConfigTFC.java:
```diff
@@ -929,6 +941,45 @@ public static void onConfigChangedEvent(ConfigChangedEvent.OnConfigChangedEvent
+	    	
+	    	@Config.Comment("Color data for summer coloring. Syntax: RED,BLUE,GREEN,ALPHA. Default=0,255,160,64")
+         	@Config.LangKey("config." + MOD_ID + ".client.grassColor.seasonColorSpring")
+         	public String seasonColorSpring = "0,255,160,64";
```
Instead of string, which requires parsing and error handling, I recommend using hexadecimal here, example:

```java
@Config.Comment("Color data for summer coloring. Syntax: ALPHA, RED, GREEN, BLUE.")
@Config.LangKey("config." + MOD_ID + ".client.grassColor.seasonColorSummer")
public int seasonColorSummer = 0x1C1E1EFF
```

This also follows ARGB standard, which iirc have some libs in java.

DisasterMoo commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,170 @@
+    	int green = Integer.parseInt(parts[2].trim());
+    	int alpha = Integer.parseInt(parts[3].trim());
+    	
+    	return new Color(red, blue, green, alpha);
```
Blue and Green are switched.

Also, If you do the above changes to the config, you may call the constructor directly:

```java
public Color(int rgba, boolean hasalpha)
```

or, if you don't care about alpha:
```java
public Color(int rgb)
```

Lumintorious committed:
Refactoring

- Changed month colors to Color array
- Changed config color strings to ARGB format
- Method resetColors() is now guarded parsing that defaults to fully transparent colors when the parsing fails
- Integrated MathHelper into the calculations
- Method getSeasonalColor() now uses a simple acces from the current month's ordinal value
- Blending functions now have weight1 be 1 - weight0

Lumintorious committed:
Alpha fix

Changed blendByAlpha to use the second color's alpha fully. If the config alpha is FF, then the grass color becomes the seasonal color entirely, 00 means it doesn't get added at all and everything in between (It was an average before and it didn't make sense)

JayJay1989 commented on src/main/java/net/dries007/tfc/client/GrassColorHandler.java:
```diff
@@ -0,0 +1,143 @@
+    	}
+    	finally
+    	{
+    		monthlyColors[Month.JULY.ordinal()] = new Color(julyCode, true);
+        	monthlyColors[Month.OCTOBER.ordinal()] = new Color(octoberCode, true);
+        	monthlyColors[Month.JANUARY.ordinal()] = new Color(januaryCode, true);
+        	monthlyColors[Month.APRIL.ordinal()] = new Color(aprilCode, true);
```
This will be overwritten by the for loop

Lumintorious replied:
It won't. The loop doesn't touch months 0, 3, 6, 9 (January, April, July and October). It just uses them to blend the months in between. For example for i = 0, only month 0 + 1 (February) and 0 + 2 (March) will be set, January and April aren't changed.
